### ISSUE TITLE
client.cleanup()

### ISSUE THREAD

**Issue by zhouhaocheng** [jborean93/pypsexec#36]
Hello Jordan! Thanks for pypsexec! When using the cleanup() method, it will report the following error.

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\Sodtware\Python27\Program\lib\site-packages\pypsexec\client.py", line 138, in cleanup
    EnumServiceState.SERVICE_STATE_ALL)
  File "D:\Sodtware\Python27\Program\lib\site-packages\pypsexec\scmr.py", line 559, in enum_services_status_w
    "display_name": (display_name + b"\x00").decode('utf-16-le'),
  File "D:\Sodtware\Python27\Program\lib\encodings\utf_16_le.py", line 16, in decode
    return codecs.utf_16_le_decode(input, errors, True)
UnicodeDecodeError: 'utf16' codec can't decode byte 0x00 in position 50: truncated data

**Comment by jborean93** [comment]
Part of cleaning up pypsexec is to enumerate all the services on your Windows host and finding the ones that start with `PAExec-` but to do that we need to enumerate all the services. It seems like your have a service installed with characters that aren't UTF-16-LE encoded or we have a bug with unpacking certain codepoints.

Would it possible to add the following to the top of your script and share the output.

```python
import logging

from pypsexec.client import Client

file_log = logging.FileHandler('pypsexec.log')
file_log.setLevel(logging.DEBUG)
file_log.setFormatter(logging.Formatter('%(asctime)s - %(name)s - %(message)s'))

logger = logging.getLogger('pypsexec.scmr')
logger.setLevel(logging.INFO)
logger.addHandler(file_log)

logger = logging.getLogger('smbprotocol.open')
logger.setLevel(logging.DEBUG)
logger.addHandler(file_log)

c = Client('server')
c.connect()
c.cleanup()
c.disconnect()
```

The file `pypsexec.log` will contain a lot of information including the raw byte read response of the `enum_service_status_w` helping us figure out why we cannot decode the raw utf-16-le values from those bytes.

**Comment by zhouhaocheng** [comment]
Hi,@jborean93,
Thank you for your answer. I have uploaded the corresponding log file. In addition, some of the display names in the service I installed contain non-English characters. Does it matter?
[pypsexec.log](https://github.com/jborean93/pypsexec/files/5685925/pypsexec.log)

**Comment by jborean93** [comment]
Sorry I read the log levels wrong and it looks like I also need the `DEBUG` logs for `pypsexec.scmr`. Would you be able to change that section to

```python
logger = logging.getLogger('pypsexec.scmr')
logger.setLevel(logging.DEBUG)
logger.addHandler(file_log)
```

**Comment by zhouhaocheng** [comment]
Hi,@jborean93,
I have re-uploaded the file, please check it out, thank you！
[pypsexec.log](https://github.com/jborean93/pypsexec/files/5686280/pypsexec.log)

**Comment by jborean93** [comment]
Thanks for that, I believe I've figured out the problem and I just need to find the proper way to fix it. The code is currently failing to decode the service with the name `Windows 内部数据库 VSS 写入程序`. The name has a raw byte value of

```
57 00 69 00 6E 00 64 00
6F 00 77 00 73 00 20 00
85 51 E8 90 70 65 6E 63
93 5E 20 00 56 00 53 00
53 00 20 00 99 51 65 51
0B 7A 8F 5E 00 00
```

When we go to extract the string from the string block we know the offset and we extract all the bytes up to the next set of `00 00` bytes. In a normal ASCII scenarios the last character will be `xx 00` so the split needs to add another byte of `00` as the split happened in between a codepoint byte representation. Unfortunately in this example the last char is `序` which when represented by UTF-16-LE is `8F 5E` breaking that faulty logic from before.  Now when we split the string block by `00 00` we have split at an actual codepoint boundary and the additional `00` byte is no longer needed. Because the byte is always added we now have a byte sequence which is no longer a valid UTF-16-LE set of codepoints and it fails with the error you have seen.

There's a quick and dirty hack I can think of implementing but I'm hoping I can avoid that for now.

**Comment by zhouhaocheng** [comment]
Hi,@jborean93,
Thank you, this service name is a Chinese name (non-English characters). Is there any solution? Or can these non-English characters be filtered out? Such non-English characters are very common on our side.

**Comment by jborean93** [comment]
Yep there's a solution I just need to implement it :). I've already got an idea of what I can do I'm just need to get CI back up and running before I put it in place. Will hopefully have a PR for this by the end of today.

**Comment by jborean93** [comment]
The changes in https://github.com/jborean93/pypsexec/pull/39 should fix your problem. If you get a chance to try them out and confirm it fixes your issue that would be great.

**Comment by zhouhaocheng** [comment]
Hi,@jborean93,
Thank you very much, cleanup() has been executed normally, but I found that ERROR_SERVICE_MARKED_FOR_DELETE error occurred when deleting some services. Although I used the following exception writing method, I found that once an exception is encountered, other normal services cannot be deleted.
from pypsexec.exception import SCMRException

try:
     c.cleanup()
except SCMRException as exc:
     if exc.return_code == 1072: # ERROR_SERVICE_MARKED_FOR_DELETE
         pass
     else:
         raise

**Comment by jborean93** [comment]
Thanks for confirming, unfortunately that bug is a longstanding issue with how PAExec works and how I integrated it into this library. I've had a project I've been working on in the background but I haven't found the time recently to work on it so it's just sitting there https://github.com/jborean93/pypsexec_service. My eventual goal is to replace the PAExec binary with my own custom service that allows me to define the behaviour and rules but that's a massive undertaking that I haven't achieved yet.

One day I will get back to it but for now it's just something we have to live with.

**Comment by jborean93** [comment]
Thanks for sharing the information and confirming the fix. I've just published [0.2.0](https://github.com/jborean93/pypsexec/releases/tag/v0.2.0) of this package which includes those changes.
